File: back/src/main/scala/models/assembly/AssemblyEvent.scala
package chrilves.kuzh.back.models.assembly

import io.circe.*
import io.circe.syntax.*
import chrilves.kuzh.back.models.Harvest

enum AssemblyEvent:
  case PublicSynchro(public: State)
  case StatusSynchro(status: State.Status)
  case PublicEvent(public: State.Event)
  case Error(reason: String, fatal: Boolean)

object AssemblyEvent:
  given assemblyEventEncoder: Encoder[AssemblyEvent] with
    final def apply(i: AssemblyEvent): Json =
      i match
        case PublicSynchro(p) =>
          Json.obj(
            "tag"   -> Json.fromString("state"),
            "state" -> p.asJson
          )
        case StatusSynchro(status) =>
          Json.obj(
            "tag"    -> Json.fromString("status"),
            "status" -> status.asJson
          )
        case PublicEvent(p) =>
          Json.obj(
            "tag"          -> Json.fromString("public_event"),
            "public_event" -> p.asJson
          )
        case Error(reason, fatal) =>
          Json.obj(
            "tag"    -> Json.fromString("error"),
            "reason" -> Json.fromString(reason),
            "fatal"  -> Json.fromBoolean(fatal)
          )

File: front/src/model/AssemblyEvent.ts
import { AssemblyState } from "./AssemblyState";
import { PublicEvent } from "./PublicEvent";

export namespace AssemblyEvent {
  export type StateW = {
    readonly tag: "state";
    state: AssemblyState;
  };

  export function stateW(state: AssemblyState): StateW {
    return {
      tag: "state",
      state: state,
    };
  }

  export type StatusW = {
    readonly tag: "status";
    status: AssemblyState.Status;
  };

  export function statusW(status: AssemblyState.Status): StatusW {
    return {
      tag: "status",
      status: status,
    };
  }

  export type PublicEventW = {
    readonly tag: "public_event";
    public_event: PublicEvent;
  };

  export function publicEvent(publicEvent: PublicEvent): PublicEventW {
    return {
      tag: "public_event",
      public_event: publicEvent,
    };
  }

  export type Error = {
    tag: "error";
    reason: string;
    fatal: boolean;
  };

  export function error(reason: string, fatal: boolean): Error {
    return {
      tag: "error",
      reason: reason,
      fatal: fatal,
    };
  }
}

export type AssemblyEvent =
  | AssemblyEvent.StateW
  | AssemblyEvent.StatusW
  | AssemblyEvent.PublicEventW
  | AssemblyEvent.Error;
